### PR TITLE
Toponaming:  Fuse call makeElementRefine; remove deprecated refineSha…

### LINKED ISSUE #15330: Empty element map if Refine is enabled
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

https://github.com/FreeCAD/FreeCAD/pull/15096#issuecomment-2200872127

```
doc = App.newDocument()
box1 = doc.addObject("Part::Box","Box")
box2 = doc.addObject("Part::Box","Box")
fuse = doc.addObject("Part::MultiFuse","Fusion")
doc.Fusion.Refine = True  # activate refinement
doc.Fusion.Shapes = [box1, box2]
doc.recompute()
print (fuse.Shape.ElementMapSize)
```
ElementMapSize is 0

### Full version info

```shell
OS: Ubuntu 22.04.4 LTS (XFCE/xubuntu)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.38090 (Git)
Build type: debug
Branch: main
Hash: fa2741fd292909eebcad965ebcdf3599bd20f25b
Python 3.10.12, Qt 5.15.3, Coin 4.0.0, Vtk 7.1.1, OCC 7.5.1
Locale: German/Germany (de_DE)
```

### Subproject(s) affected?

Part

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by bgbsww**
…peIfActive  fix #15330

### CODE REVIEW

**Comment by CalligaroV** [comment]
@bgbsww, looks like https://github.com/FreeCAD/FreeCAD/pull/15897/commits/24f507fd44fbf44cb6f37f8c7e704959393161b2 is making https://github.com/FreeCAD/FreeCAD/blob/87e73967da5734e100d5ac4b2fd5af1648a02572/src/Mod/Part/parttests/ColorPerFaceTest.py#L150 fail.
(My fault, sorry!)

Can you revert that modification?
If you want I can do it myself but if so I'll have to bother you with another PR on your fork :disappointed_relieved:

File: tests/src/Mod/Part/App/FeaturePartFuse.cpp [hunk 1]
```diff
@@ -187,13 +187,25 @@ TEST_F(FeaturePartFuseTest, testRefine)
         ts.getSubTopoShapes(TopAbs_FACE);  // TopAbs_WIRE alternate approach
     // Assert two boxes, plus redundant faces at the joint.
     EXPECT_EQ(subs.size(), 14);
+    // 14 Faces
+    // 28 Edges
+    // 16 Vetexes
```
Comment (chennes) [inline tests/src/Mod/Part/App/FeaturePartFuse.cpp:192]:
```suggestion
    // 16 Vertices
```

File: tests/src/Mod/Part/App/FeaturePartFuse.cpp [hunk 2]
```diff
@@ -187,13 +187,25 @@ TEST_F(FeaturePartFuseTest, testRefine)
...
     // Act
     _fuse->Refine.setValue(true);
     _fuse->execute();
     ts = _fuse->Shape.getValue();
     subs = ts.getSubTopoShapes(TopAbs_FACE);
     // Assert we now just have one big box
     EXPECT_EQ(subs.size(), 6);
+    // 6 Faces
+    // 12 Edges
+    // 8 Vetexes
```
Comment (chennes) [inline tests/src/Mod/Part/App/FeaturePartFuse.cpp:205]:
```suggestion
    // 8 Vertices
```

**Review by chennes — APPROVED** [review]
Two spelling errors I will fix on merging

**Comment by wwmayer** [comment]
This PR causes a regression.
1. Create a Part box
2. Create a Part cylinder
3. Select both and set random colours
4. Merge them => Crash

In case it doesn't crash yet then set the Refine property of the fuse object to True and recompute the document.

Before this PR the crash did not occur.

**Comment by wwmayer** [comment]
The test code of https://github.com/FreeCAD/FreeCAD/issues/15330 should be added to a unit test.

**Comment by bgbsww** [comment]
Yes, makeElementRefine is simply wrong.  Adjusting to fix, including recommended test + more.

**Comment by CalligaroV** [comment]
@wwmayer can an update to https://github.com/FreeCAD/FreeCAD/blob/0dae7353d1eb38097fe3cd520fc2e9871dc85559/tests/src/Mod/Part/App/TopoShapeMakeElementRefine.cpp#L28 be enough as unit tests for #15330 or do you want a dedicated python test ? (haven't checked yet if there's already one)

In `TopoShapeMakeElementRefine.cpp` there will be expected values for a `Part::Fuse` and the refined shape (I already updated the values in a branch I was about to push)

**Comment by bgbsww** [comment]
Update to that test doesn't hurt, but does it test the MultiFuse shown in Werner's test?  I've added a variant of that to the TestTopologicalNamingProblem.py tests.

**Comment by bgbsww** [comment]
Update:  This doesn't look like a direct error, but an interaction with the new materials system.  The crash is called when you refine() on shapes with different materials ( including if one is just default ) which is why the randomColor setting is needed to trigger it.

It does work before this PR, but the difference here is supposed to be making the elementMaps not get erased accidentally on a fuse call.

**Comment by bgbsww** [comment]
Duh, because we now fail to assemble a ShapeHistory which breaks materials.  Okay, I see it now.

**Pull request merged**